Moller-Plesset/MP2_Gradient.py

```diff
@@ -0,0 +1,506 @@
+Pabic = 2 * Ppqrs[nocc:,nocc:,:nocc,nocc:] - Ppqrs[nocc:,nocc:,:nocc,nocc:].swapaxes(0,1)
+Pabci = 2 * Ppqrs[nocc:,nocc:,nocc:,:nocc] - Ppqrs[nocc:,nocc:,nocc:,:nocc].swapaxes(0,1)
+Ppqrs[nocc:,nocc:,nocc:,:nocc] = Pabci
+
+
+# <VV|VV>
+# <--- sum_abcd ( [2 P(a,b,c,d) - P(a,b,d,c)] * <ab|cd> )
+Pabic = 2 * Ppqrs[nocc:,nocc:,:nocc,nocc:] - Ppqrs[nocc:,nocc:,:nocc,nocc:].swapaxes(0,1)
+Pabcd = 2 * Ppqrs[nocc:,nocc:,nocc:,nocc:] - Ppqrs[nocc:,nocc:,nocc:,nocc:].swapaxes(2,3)
+Ppqrs[nocc:,nocc:,nocc:,nocc:] = Pabcd
+
+
+
+Gradient = {}
+Gradient["N"] = np.zeros((natoms, 3))
+Gradient["S"] = np.zeros((natoms, 3))
+Gradient["V"] = np.zeros((natoms, 3))
+Gradient["T"] = np.zeros((natoms, 3))
+Gradient["J"] = np.zeros((natoms, 3))
+Gradient["K"] = np.zeros((natoms, 3))
+Gradient["Total"] = np.zeros((natoms, 3))
+
+# 1st Derivative of Nuclear Repulsion
+Gradient["N"] = psi4.core.Matrix.to_array(mol.nuclear_repulsion_energy_deriv1([0,0,0]))
+
+psi4.core.print_out("\n\n")
+N_grad = psi4.core.Matrix.from_array(Gradient["N"])
+N_grad.name = "NUCLEAR GRADIENT"
+N_grad.print_out()
+
+# Build Integral Gradients
+cart = ['_X', '_Y', '_Z']
+oei_dict = {"S" : "OVERLAP", "T" : "KINETIC", "V" : "POTENTIAL"}
+deriv1_mat = {}
+deriv1_np = {}
+
+# 1st Derivative of OEIs
+for atom in range(natoms):
+    for key in  oei_dict:
+        deriv1_mat[key + str(atom)] = mints.mo_oei_deriv1(oei_dict[key], atom, C, C)
+        for p in range(3):
+            map_key = key + str(atom) + cart[p]
+            deriv1_np[map_key] = np.asarray(deriv1_mat[key + str(atom)][p])
+            if key == "S":
+                Gradient[key][atom, p] = 2.0 * np.einsum('pq,pq->', I, deriv1_np[map_key])
+            else:
+                Gradient[key][atom, p] = np.einsum("pq,pq->", ref_opdm + (2 * Ppq), deriv1_np[map_key])
+
+psi4.core.print_out("\n\n OEI Gradients:\n\n")
+for key in Gradient:
+    Mat = psi4.core.Matrix.from_array(Gradient[key])
+    if key in oei_dict:
+        Mat.name = oei_dict[key] + " GRADIENT"
+        Mat.print_out()
+        psi4.core.print_out("\n")
+
+
+# 1st Derivative of TEIs
+for atom in range(natoms):
+    string = "TEI" + str(atom)
+    deriv1_mat[string] = mints.mo_tei_deriv1(atom, C, C, C, C)
+    for p in range(3):
+        map_key = string + cart[p]
+        deriv1_np[map_key] = np.asarray(deriv1_mat[string][p])
+
+        # Reference OPDM component of TEI gradients
+        Gradient["J"][atom, p] += 2.0 * np.einsum('pq,pqmm->', ref_opdm, deriv1_np[map_key][:,:,:nocc,:nocc])
+        Gradient["K"][atom, p] -= 1.0 * np.einsum('pq,pmmq->', ref_opdm, deriv1_np[map_key][:,:nocc,:nocc,:])
+
+        # Reference TPDM component of TEI gradients
+        Gradient["J"][atom, p] += 4.0 * np.einsum('pqrs,prqs->', ref_tpdm, deriv1_np[map_key])
+
+        # Contract MP2 TPDM altogether
+        Gradient["J"][atom, p] += 4.0 * np.einsum('pqrs,prqs->', Ppqrs, deriv1_np[map_key])
+
+        # Contract MP2 TPDM by individual components
```

Comment:
If these commented blocks are potentially useful then leave them, but if not you can remove them. May want to see what @dgasmith, @dsirianni have to say

Reply:
Yeah, these might be helpful to me in the future when I'm working on things on Psi4 side of things, where I'll have to implement each individual block of the TPDM separately. But I imagine I could probably get rid of them here.